libraries/wasm/src/doc.rs

```diff
@@ -0,0 +1,260 @@
+            Ok(diff) => self.0.verify_diff(&diff).is_ok(),
+            Err(_) => false,
+        }
+    }
+
+    #[wasm_bindgen]
+    pub fn update_service(&mut self, did: WasmDID, url: String, service_type: String) -> Result<Doc, JsValue> {
+        let endpoint = ServiceEndpoint::Url(url.parse().map_err(js_err)?);
+        let service = ServiceBuilder::default()
+            .id(DID::parse(did.0).map_err(js_err)?)
+            .service_type(service_type)
+            .endpoint(endpoint)
+            .build()
+            .map_err(js_err)?;
+        self.0.update_service(service);
+        Ok(Doc(self.0.clone()))
+    }
+
+    #[wasm_bindgen]
+    pub fn clear_services(&mut self) -> Result<Doc, JsValue> {
+        self.0.clear_services();
+        Ok(Doc(self.0.clone()))
+    }
+
+    #[wasm_bindgen]
+    pub fn update_public_key(&mut self, public_key: &PubKey) -> Result<Doc, JsValue> {
+        self.0.update_public_key(public_key.0.clone());
+        Ok(Doc(self.0.clone()))
+    }
+
+    #[wasm_bindgen]
+    pub fn update_auth(&mut self, public_key: &PubKey) -> Result<Doc, JsValue> {
+        self.0.update_auth(public_key.0.clone());
+        Ok(Doc(self.0.clone()))
+    }
+
+    #[wasm_bindgen]
+    pub fn update_assert(&mut self, public_key: &PubKey) -> Result<Doc, JsValue> {
+        self.0.update_assert(public_key.0.clone());
+        Ok(Doc(self.0.clone()))
+    }
+
+    #[wasm_bindgen]
+    pub fn update_verification(&mut self, public_key: &PubKey) -> Result<Doc, JsValue> {
+        self.0.update_verification(public_key.0.clone());
+        Ok(Doc(self.0.clone()))
+    }
+
+    #[wasm_bindgen]
+    pub fn update_delegation(&mut self, public_key: &PubKey) -> Result<Doc, JsValue> {
+        self.0.update_delegation(public_key.0.clone());
+        Ok(Doc(self.0.clone()))
+    }
+
+    #[wasm_bindgen]
+    pub fn update_invocation(&mut self, public_key: &PubKey) -> Result<Doc, JsValue> {
+        self.0.update_invocation(public_key.0.clone());
+        Ok(Doc(self.0.clone()))
+    }
+
+    #[wasm_bindgen]
+    pub fn update_agreement(&mut self, public_key: &PubKey) -> Result<Doc, JsValue> {
+        self.0.update_agreement(public_key.0.clone());
+        Ok(Doc(self.0.clone()))
+    }
+
+    #[wasm_bindgen]
+    pub fn update_time(&mut self) {
+        self.0.update_time();
+    }
+
+    #[wasm_bindgen]
+    pub fn resolve_key(&mut self, key_relation: &str) -> Result<PubKey, JsValue> {
+        let relation = match key_relation {
+            "VerificationMethod" => KeyRelation::VerificationMethod,
+            "Authentication" => KeyRelation::Authentication,
+            "AssertionMethod" => KeyRelation::AssertionMethod,
+            "KeyAgreement" => KeyRelation::KeyAgreement,
+            "CapabilityInvocation" => KeyRelation::CapabilityInvocation,
+            "CapabilityDelegation" => KeyRelation::CapabilityDelegation,
+            _ => panic!("Unknown KeyRelation"),
+        };
+
+        match self.0.resolve_key(0, relation) {
```

Comment:
you can use `expect` to "unwrap or panic" with a message

`self.0.resolve_key(0, relation).expect("Key not found")`